node/src/exchange/client.rs

```diff
@@ -56,63 +96,821 @@ impl ExchangeClientHandler {
+                return;
+            }
+
+            // Count peers per response
+            for resp in &resps {
+                *counter.entry(resp.hash()).or_default() += 1;
+            }
+
+            // Sort by height and then peers in descending order
+            resps.sort_unstable_by_key(|resp| {
+                let num_of_peers = counter[&resp.hash()];
+                Reverse((resp.height(), num_of_peers))
+            });
+
+            // Return the header with the highest height that was received by at least 2 peers
+            for resp in &resps {
+                if counter[&resp.hash()] >= MIN_HEAD_RESPONSES {
+                    respond_to.maybe_send_ok(vec![resp.to_owned()]);
+                    return;
+                }
+            }
+
+            // Otherwise return the header with the maximum height
+            respond_to.maybe_send_ok(vec![resps[0].to_owned()]);
+        });
     }
 
     #[instrument(level = "trace", skip(self, responses), fields(responses.len = responses.len()))]
     pub(super) fn on_response_received(
         &mut self,
-        request_id: RequestId,
+        peer: PeerId,
+        request_id: S::RequestId,
         responses: Vec<HeaderResponse>,
     ) {
-        let Some(ReqInfo { amount, respond_to }) = self.reqs.remove(&request_id) else {
+        let Some(state) = self.reqs.remove(&request_id) else {
             return;
         };
 
-        trace!("Response received. Expected amount = {amount}");
+        trace!(
+            "Response received. Expected amount = {}",
+            state.request.amount
+        );
+
+        match self.decode_and_verify_responses(&state.request, &responses) {
+            Ok(headers) => {
+                // TODO: Increase peer score
+                state.respond_to.maybe_send_ok(headers);
+            }
+            Err(e) => {
+                // TODO: Decrease peer score
+                state.respond_to.maybe_send_err(e);
+            }
+        }
+    }
+
+    fn decode_and_verify_responses(
+        &mut self,
+        request: &HeaderRequest,
+        responses: &[HeaderResponse],
+    ) -> Result<Vec<ExtendedHeader>, ExchangeError> {
+        let amount = usize::try_from(request.amount).expect("validated in send_request");
 
         if responses.len() != amount {
-            // TODO: should we define a separate error for this case?
-            respond_to.maybe_send_err(ExchangeError::InvalidResponse);
-            return;
+            // When server returns an error, it encodes it as one HeaderResponse.
+            // In that case propagate the decoded error.
+            if responses.len() == 1 {
+                responses[0].to_extended_header()?;
+            }
+
+            // Otherwise report it as InvalidResponse
+            //
+            // TODO: Should we relax this? Server can return a sub-range.
+            return Err(ExchangeError::InvalidResponse);
         }
```

Comment:
This is wrong and should be relaxed. This is what Celestia is doing:

1. Send request for the range (e.g: start = 10, amount = 10)
2. Receives a sub-range response (e.g. [10, 14])
3. Accepts the partial response and submits a new one for the remaining range (e.g. start = 15, amount = 5)

Reply:
isn't it what we should do in https://github.com/eigerco/celestia-node-rs/issues/62?

Reply:
Yes. I relaxed the check for now and a partial range will be returned. If you prefer the more strict version let me know.

Reply:
Actually I changed my mind. I reverted the relaxed version because it changes the behavior and may affect other logic later on. We should guarantee that response of `get_range_*` should be the whole range or an error.